llvm/lib/CodeGen/CodeGenPrepare.cpp

```diff
@@ -7387,17 +7387,38 @@ bool CodeGenPrepare::optimizeSwitchType(SwitchInst *SI) {
   // attribute. In that case, we can avoid an unnecessary mask/extension by
   // matching the argument extension instead.
   Instruction::CastOps ExtType = Instruction::ZExt;
-  // Some targets prefer SExt over ZExt.
-  if (TLI->isSExtCheaperThanZExt(OldVT, RegType))
-    ExtType = Instruction::SExt;
 
+  unsigned SExtUsers = 0;
+  unsigned ZExtUsers = 0;
   if (auto *Arg = dyn_cast<Argument>(Cond)) {
     if (Arg->hasSExtAttr())
-      ExtType = Instruction::SExt;
+      SExtUsers++;
     if (Arg->hasZExtAttr())
-      ExtType = Instruction::ZExt;
+      ZExtUsers++;
+  }
+  for (auto U : Cond->users()) {
+    if (isa<SExtInst>(U)) {
+      SExtUsers++;
+    } else if (isa<ZExtInst>(U)) {
+      ZExtUsers++;
+    } else if (auto *IC = dyn_cast<ICmpInst>(U)) {
+      if (IC->isSigned()) {
+        SExtUsers++;
+      } else if (IC->isUnsigned()) {
+        ZExtUsers++;
+      }
+    }
+  }
+  if (SExtUsers > ZExtUsers) {
+    ExtType = Instruction::SExt;
+  } else {
+    ExtType = Instruction::ZExt;
   }
 
+  // Some targets prefer SExt over ZExt.
+  if (TLI->isSExtCheaperThanZExt(OldVT, RegType))
```

Comment:
What was the point of doing this user scan if you just throw away the result?

Reply:
Because in some cases we have more ZExtUsers but SExt is cheaper so if create a ZExt we add more instructions. 
I checked spec2017 for RISC-V with the precedence with this and It was doing worse on 500.perlbench_r if we check SExtIsCheaperThanZExt before the user scan. I can share the results later

Reply:
This is the text sizes with SExtIsCheaperThanZExt take no precedence over the user scan 
![SextIsCheaperThanZextisBeforeUserScan](https://github.com/llvm/llvm-project/assets/45971450/3608815f-28d3-483d-be3a-21d61b0340c9)
And this is the text sizes with SExtIsCheaperThanZExt taking precedence than user scan.
![SextIsCheaperThanZextisAfterUserScan](https://github.com/llvm/llvm-project/assets/45971450/d4809dba-67f9-4087-84da-0c4121030d62)

Reply:
I don't mean the end result, I mean the computation you perform in this code is discarded based on isSExtCheaperThanZExt

Reply:
Yes isSExtCheaperThanZExt needs to take precedence if it doesn't agree with the computation that I perform otherwise the code size increases as I showed above.
Maybe the gain is very small compared to the cost because we need a User Scan.

Reply:
My reading of @arsenm's comment is that maybe it would be better to avoid the user scan if we dont need it as such:
```
if (TLI->isSExtCheaperThanZExt(OldVT, RegType))
  ExtType = Instruction::SExt;
else {
 // User scan above
}

Reply:
If one of the users is a ZExt, then ZExt is preferable to SExt even if SExt is cheaper.